src/v/kafka/server/errors.h

```diff
@@ -36,6 +36,8 @@ constexpr error_code map_topic_error_code(cluster::errc code) {
         return error_code::request_timed_out;
     case cluster::errc::invalid_topic_name:
         return error_code::invalid_topic_exception;
+    case cluster::errc::no_leader_controller:
+        return error_code::leader_not_available;
```

Comment:
there is also the error codes `not_controller` and `stale_controller_epoch` which seem very similar but more specific to the situation. can you comment on these in relation to the leader not available error? maybe @twmb has some insight into what makes sense. apologies if this has already been hashed out!

Reply:
Is this being returned to a client? StaleControllerEpoch is not retriable, NotController is.

For a client, NotController allows the client to know it needs to refresh metadata and then retry. LeaderNotAvailable signals to the client that the leader is the same, but it's down. The client just backs off and retries (no metadata refresh).

Reply:
I guess the latter is preferred then, as we have no idea which broker will become leader.

Reply:
Is there a metadata refresh on LeaderNotAvailable?

Reply:
There could be but there doesn't have to be. I don't retry on that error in franz-go, for example -- all I do is back off.

NotController is basically the broker saying "your current state is wrong", LeaderNotAvailable is basically saying "you're right, but I can't handle this". In a way, similar to 4xx vs 5xx.

Reply:
FWIW looking at librdkafka, it loads metadata on either error.

Reply:
I would vote for `NotController`

Reply:
Done

Reply:
> Is this being returned to a client? StaleControllerEpoch is not retriable, NotController is. (edit: also StaleControllerEpoch is a broker<=>broker error, not something that clients should see).
> 
> For a client, NotController allows the client to know it needs to refresh metadata and then retry. LeaderNotAvailable signals to the client that the leader is the same, but it's down. The client just backs off and retries (no metadata refresh).

Thanks @twmb this explanation was great